3 Module Development Exercise - Recursive Unhider/recursiveUnhider-rt/src/com/tridiumuniversity/recursiveUnhider/BUnhider.java

```diff
@@ -0,0 +1,124 @@
+
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+
+@NiagaraType
+@NiagaraProperty(
+        name = "unhideRegex",
+        type = "String",
+        defaultValue = "new String()",
+        flags = Flags.SUMMARY
+)
+@NiagaraAction(
+        name = "unhideSlots",
+        flags = Flags.SUMMARY | Flags.ASYNC
+)
+public class BUnhider extends BComponent
+{
+//region /*+ ------------ BEGIN BAJA AUTO GENERATED CODE ------------ +*/
+//@formatter:off
+/*@ $com.tridiumuniversity.recursiveUnhider.BUnhider(2130432398)1.0$ @*/
+/* Generated Mon Nov 20 12:07:19 EST 2023 by Slot-o-Matic (c) Tridium, Inc. 2012-2023 */
+
+  //region Property "unhideRegex"
+
+  /**
+   * Slot for the {@code unhideRegex} property.
+   * @see #getUnhideRegex
+   * @see #setUnhideRegex
+   */
+  public static final Property unhideRegex = newProperty(Flags.SUMMARY, new String(), null);
+
+  /**
+   * Get the {@code unhideRegex} property.
+   * @see #unhideRegex
+   */
+  public String getUnhideRegex() { return getString(unhideRegex); }
+
+  /**
+   * Set the {@code unhideRegex} property.
+   * @see #unhideRegex
+   */
+  public void setUnhideRegex(String v) { setString(unhideRegex, v, null); }
+
+  //endregion Property "unhideRegex"
+
+  //region Action "unhideSlots"
+
+  /**
+   * Slot for the {@code unhideSlots} action.
+   * @see #unhideSlots()
+   */
+  public static final Action unhideSlots = newAction(Flags.SUMMARY | Flags.ASYNC, null);
+
+  /**
+   * Invoke the {@code unhideSlots} action.
+   * @see #unhideSlots
+   */
+  public void unhideSlots() { invoke(unhideSlots, null, null); }
+
+  //endregion Action "unhideSlots"
+
+  //region Type
+
+  @Override
+  public Type getType() { return TYPE; }
+  public static final Type TYPE = Sys.loadType(BUnhider.class);
+
+  //endregion Type
+
+//@formatter:on
+//endregion /*+ ------------ END BAJA AUTO GENERATED CODE -------------- +*/
+
+    public void doUnhideSlots()
+    {
+        slotPattern = Pattern.compile(getUnhideRegex());
+        recurse((BComponent) getParent());
+    }
+
+    private void recurse(BComponent parent)
+    {
+        BValue child;
+        for (Slot slot : parent.getSlotsArray())
+        {
+            if (slot.isProperty())
+            {
+                child = parent.get(slot.asProperty());
+                if (child instanceof BComponent) recurse((BComponent) child);
+            }
+            if (Flags.isHidden(parent, slot) && slotPattern.matcher(slot.getName()).matches())
+            {
+                Flags.remove(parent, slot, null, Flags.HIDDEN);
+            }
+            if (Flags.isHidden(parent, slot)) System.out.println("Hidden slot name did not match");
+            else System.out.println("Unhid a slot");
+        }
+    }
+
+    @Override
+    public void started()
+    {
+        exec = newSingleThreadExecutor();
+    }
+
+    @Override
+    public IFuture post(Action a, BValue arg, Context c)
+    {
+        if (unhideSlots.equals(a))
```

Comment:
Since you're only using a single thread executor, you might not run into problems with this. In general, however, when you're multithreading like this, you probably want to check here if the action is currently running and, if so, to ignore future invocations of the action until the first invocation completes. You'd run into a problem if someone sets `unhideRegex` to `abc`, invokes this action, then (while the first invocation is still running) sets `unhideRegex` to `def` and invokes the action again. The `Pattern` field would change as a result of the second invocation, which would change the slots that are being unhidden in the thread running the first invocation.

The easiest way to guard against this is to use an `AtomicBoolean`. You can set it to `true` in this method before calling `exec.execute`, then set it to `false` at the end of `doUnhideSlots`. Then you just make sure your `AtomicBoolean` is `false` before calling `exec.execute`.